File: shopify-order-creator/ts/src/verify/shipments.ts
/** Allocation-state checks against staging-shipments ITEM# rows. Ports regression/verify/shipments.py. */

import type { AllocationSummary, ShipmentItem } from "../readers/dynamoReader";
import { VerificationError } from "./index";

function mapsEqual(a: Record<string, number>, b: Record<string, number>): boolean {
  const aKeys = Object.keys(a);
  const bKeys = Object.keys(b);
  return aKeys.length === bKeys.length && aKeys.every((key) => a[key] === b[key]);
}

/**
 * One ITEM# row per unit: total shipment units per SKU must equal the
 * ordered quantity (DynamoDB does NOT merge duplicates like Shopify does).
 */
export function assertUnitCounts(summary: AllocationSummary, expectedSkus: Record<string, number>, orderName: string): void {
  if (!mapsEqual(summary.skuUnits, expectedSkus)) {
    throw new VerificationError(
      "shipments.unit_counts",
      expectedSkus,
      summary.skuUnits,
      `order ${orderName}; by_store=${JSON.stringify(summary.byStore)}`,
    );
  }
}

/**
 * Each SKU landed where the seed plan forced it.
 *
 * expectedAllocation: {sku: storeNumberOrUNDELIVERABLE}, e.g.
 *   {"32625134": "100"} or {"33413679": "UNDELIVERABLE"}.
 */
export function assertAllocation(
  summary: AllocationSummary,
  expectedAllocation: Record<string, string>,
  orderName: string,
): void {
  if (summary.unallocated > 0) {
    throw new VerificationError(
      "shipments.allocated",
      "all items allocated or undeliverable",
      `${summary.unallocated} item(s) with no store value`,
      `order ${orderName}; by_store=${JSON.stringify(summary.byStore)}`,
    );
  }

  const actual: Record<string, Set<string>> = {};
  for (const [store, skus] of Object.entries(summary.byStore)) {
    for (const sku of skus) {
      if (!actual[sku]) {
        actual[sku] = new Set();
      }
      actual[sku].add(store);
    }
  }

  for (const [sku, expectedStore] of Object.entries(expectedAllocation)) {
    const stores = actual[sku];
    if (!stores) {
      throw new VerificationError(
        "shipments.allocation",
        { [sku]: expectedStore },
        { [sku]: null },
        `order ${orderName}: no ITEM# rows for sku`,
      );
    }
    const storesArray = Array.from(stores).sort();
    if (storesArray.length !== 1 || storesArray[0] !== String(expectedStore)) {
      throw new VerificationError(
        "shipments.allocation",
        { [sku]: String(expectedStore) },
        { [sku]: storesArray },
        `order ${orderName}`,
      );
    }
  }
}

/** After undeliverable cleanup, the refunded SKUs have no ITEM# rows left. */
export function assertItemsRemoved(shipmentItems: ShipmentItem[], removedSkus: string[], orderName: string): void {
  const remaining = Array.from(new Set(shipmentItems.filter((item) => removedSkus.includes(item.sku)).map((item) => item.sku))).sort();
  if (remaining.length > 0) {
    throw new VerificationError(
      "shipments.cleanup",
      `no rows for ${JSON.stringify([...removedSkus].sort())}`,
      `rows still present for ${JSON.stringify(remaining)}`,
      `order ${orderName}`,
    );
  }
}

File: shopify-order-creator/regression/verify/shipments.py
"""Allocation-state checks against staging-shipments ITEM# rows."""

from regression.verify import VerificationError


def assert_unit_counts(summary: dict, expected_skus: dict[str, int], order_name: str):
    """
    One ITEM# row per unit: total shipment units per SKU must equal the
    ordered quantity (DynamoDB does NOT merge duplicates like Shopify does).
    """
    if summary["sku_units"] != expected_skus:
        raise VerificationError(
            "shipments.unit_counts", expected_skus, summary["sku_units"],
            detail=f"order {order_name}; by_store={summary['by_store']}",
        )


def assert_allocation(summary: dict, expected_allocation: dict[str, str], order_name: str):
    """
    Each SKU landed where the seed plan forced it.

    expected_allocation: {sku: store_number_or_UNDELIVERABLE}, e.g.
        {"32625134": "100"} or {"33413679": "UNDELIVERABLE"}.
    """
    if summary["unallocated"]:
        raise VerificationError(
            "shipments.allocated", "all items allocated or undeliverable",
            f"{summary['unallocated']} item(s) with no store value",
            detail=f"order {order_name}; by_store={summary['by_store']}",
        )

    # Build actual sku -> set of stores from the by_store map.
    actual: dict[str, set] = {}
    for store, skus in summary["by_store"].items():
        for sku in skus:
            actual.setdefault(sku, set()).add(str(store))

    for sku, expected_store in expected_allocation.items():
        stores = actual.get(sku)
        if not stores:
            raise VerificationError(
                "shipments.allocation", {sku: expected_store}, {sku: None},
                detail=f"order {order_name}: no ITEM# rows for sku",
            )
        if stores != {str(expected_store)}:
            raise VerificationError(
                "shipments.allocation", {sku: str(expected_store)}, {sku: sorted(stores)},
                detail=f"order {order_name}",
            )


def assert_items_removed(shipment_items: list[dict], removed_skus: list[str], order_name: str):
    """
    After undeliverable cleanup, the refunded SKUs' ITEM# rows are removed.

    LIVE FINDING (Jul 17, TS runs #9706/#9707): staging-shipments rows are
    never physically deleted — cleanup flips the row status to "REMOVED".
    Accept either absence or status == "REMOVED". (TS harness has the same
    fix; keep both in sync.)
    """
    remaining = sorted({
        i["sku"] for i in shipment_items
        if i["sku"] in removed_skus and str(i.get("status", "")).upper() != "REMOVED"
    })
    if remaining:
        raise VerificationError(
            "shipments.cleanup",
            f"rows absent or status REMOVED for {sorted(removed_skus)}",
            f"active rows still present for {remaining}",
            detail=f"order {order_name}",
        )
